### ISSUE TITLE
single_space_after_construct removing linebreaks between constructs and opening braces

### ISSUE THREAD

**Issue by Andrew-Staves-Activ** [FriendsOfPHP/PHP-CS-Fixer#5506]
## Bug report

   * PHP version: 7.2.34
   * PHP CS Fixer version: 2.18.2

The `single_space_after_construct` fixer is removing linebreaks (and indentation), causing braces to be repositioned. I think this is only relevant to `do`, `else`, `finally`, and `try`.

### Configuration file
```php
<?php

$finder = PhpCsFixer\Finder::create()->in(__DIR__);

return PhpCsFixer\Config::create()
	->setFinder($finder)
	->setIndent("\t")
	->setLineEnding("\n")
	->setRules([
		"single_space_after_construct" => true,
	])
;
```

### Code snippet that reproduces the problem

```php
<?php

class Foo
{
	public  function bar($baz)
	{
		if ($baz == 900)
		{
			try
			{
				echo  "Hello!";
			}
			catch(\Throwable $thrown)
			{
				echo "Error: {$thrown->getMessage()}";
			}
			finally
			{
				echo "Moving on.";
			}
		}

		if ($baz == true)
		{
			echo"Why?";
		}
		else
		{
			echo "Ha?";
		}
	}
}
```

### Desired/expected output
```php
<?php

class Foo
{
	public function bar($baz)
	{
		if ($baz == 900)
		{
			try
			{
				echo "Hello!";
			}
			catch (\Throwable $thrown)
			{
				echo "Error: {$thrown->getMessage()}";
			}
			finally
			{
				echo "Moving on.";
			}
		}

		if ($baz == true)
		{
			echo "Why?";
		}
		else
		{
			echo "Ha?";
		}
	}
}
```

### Actual output
```php
<?php

class Foo
{
	public function bar($baz)
	{
		if ($baz == 900)
		{
			try {
				echo "Hello!";
			}
			catch (\Throwable $thrown)
			{
				echo "Error: {$thrown->getMessage()}";
			}
			finally {
				echo "Moving on.";
			}
		}

		if ($baz == true)
		{
			echo "Why?";
		}
		else {
			echo "Ha?";
		}
	}
}
```

### Workaround (adjusted config file)
```php
<?php

$finder = PhpCsFixer\Finder::create()->in(__DIR__);

return PhpCsFixer\Config::create()
	->setFinder($finder)
	->setIndent("\t")
	->setLineEnding("\n")
	->setRules([
		"single_space_after_construct" => [
			"constructs" => [
				"abstract",
				"as",
				"attribute",
				"break",
				"case",
				"catch",
				"class",
				"clone",
				"comment",
				"const",
				"const_import",
				"continue",
				// "do",
				"echo",
				// "else",
				"elseif",
				"extends",
				"final",
				// "finally",
				"for",
				"foreach",
				"function",
				"function_import",
				"global",
				"goto",
				"if",
				"implements",
				"include",
				"include_once",
				"instanceof",
				"insteadof",
				"interface",
				"match",
				"named_argument",
				"new",
				"open_tag_with_echo",
				"php_doc",
				"php_open",
				"print",
				"private",
				"protected",
				"public",
				"require",
				"require_once",
				"return",
				"static",
				"throw",
				"trait",
				// "try",
				"use",
				"use_lambda",
				"use_trait",
				"var",
				"while",
				"yield",
				"yield_from",
			],
		],
	])
;
```

Thanks!

**Comment by Wimble84** [comment]
Same problem on 2.18.4

**Comment by kubawerlos** [comment]
@Andrew-Staves-Activ @Wimble84 what problem are you exactly reporting?

Rule [`single_space_after_construct`](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.18.4/doc/rules/language_construct/single_space_after_construct.rst) is suppose to remove those linebreaks.

**Comment by Wimble84** [comment]
Line break are not single space. If you look at outputs given by @Andrew-Staves-Activ you can see that line breaks are removed on **try**, **finally** and **if** but NOT on **catch** and **else**. Why ? All these keywords are in the list. 

This behaviour breaks the rule **position_after_control_structures => 'next'**

**Comment by Andrew-Staves-Activ** [comment]
What @Wimble84 said.

**Comment by kubawerlos** [comment]
@Wimble84 

> line breaks are removed on try, finally and else but NOT on catch and if. Why ?

Well, `catch` and `if` do not have line breaks after them.

Don't look what is removed or not, but what is added:
 - break line after construct -> changed to single space
 - no whitespace after construct -> added single space

The fixer is making all the constructs configured to have single space after them - either by replacing whitespace not being single space or by adding one.

> This behaviour breaks the rule position_after_control_structures => 'next'

There is a test to cover exactly that scenario: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.18/tests/Fixtures/Integration/priority/single_space_after_construct%2Cbraces.test

@Andrew-Staves-Activ 

> That's news to me

Let me border the key word on that sentence:

> Rule single_space_after_construct is suppose to remove *those* linebreaks.

By "those" I meant the linebreaks that are after constructs that suppose to have single space after them, so the fixer does it job - replaces the linebreaks with single space.

> From the current documentation, I expected the fixer to only affect the line with the construct

What gave you that impression? Descriptions says "Ensures a single space after language constructs", nothing about any exception, but there is no example with line break and 3 without it - did that give you that impression? I think this is something we could/should improve.

> this seems more something that the "braces" fixer should do

For braces yes, but many times many of the constructs from the list are not followed by any braces.

> the behaviour is not consistent for different constructs

I'd say it is - after fix all the constructs have single space after them.

**Comment by Andrew-Staves-Activ** [comment]
> there is no example with line break and 3 without it - did that give you that impression?

Pretty much that, and also there being no mention in the accompanying text.

> > this seems more something that the "braces" fixer should do

> For braces yes, but many times many of the constructs from the list are not followed by any braces.

>> the behaviour is not consistent for different constructs

> I'd say it is - after fix all the constructs have single space after them.

Ok, I can see what you mean, thanks for explaining. I would like to use this in a way that does not lead to inconsistent blocks, in other words, I don't want:
```php
<?php

if ($baz == true)
{
	echo "Why?";
}
else {
	echo "Ha?";
}
```
I want:
```php
<?php

if ($baz == true)
{
	echo "Why?";
}
else
{
	echo "Ha?";
}
```
From what you've said, I can understand why this happens with just the `single_space_after_construct` fixer. So, if I try to work around this with the following configuration (_defaults commented for reference_):
```php
<?php

$finder = PhpCsFixer\Finder::create()->in(__DIR__);

return PhpCsFixer\Config::create()
	->setFinder($finder)
	->setIndent("\t")
	->setLineEnding("\n")
	->setRules([
		"braces" => [
			// "allow_single_line_anonymous_class_with_empty_body" => false,
			// "allow_single_line_closure" => false,
			// "position_after_anonymous_constructs" => "same",
			"position_after_control_structures" => "next",
			// "position_after_functions_and_oop_constructs" => "next",
		],
		"single_space_after_construct" => true,
	])
;
```
This input:
```php
<?php

class Foo
{
	public  function bar($baz)
	{
		if ($baz == 900)
		{
			try
			{
				echo  "Hello!";
			}
			catch(\Throwable $thrown)
			{
				echo "Error: {$thrown->getMessage()}";
			}
			finally
			{
				echo "Moving on.";
			}
		}

		if ($baz == true)
		{
			echo"Why?";
		}
		else
		{
			echo "Ha?";
		}
	}
}
```
Results in this output:
```php
<?php

class Foo
{
	public function bar($baz)
	{
		if ($baz == 900)
		{
			try {
				echo "Hello!";
			}
			catch (\Throwable $thrown)
			{
				echo "Error: {$thrown->getMessage()}";
			}
			finally {
				echo "Moving on.";
			}
		}

		if ($baz == true)
		{
			echo "Why?";
		}
		else {
			echo "Ha?";
		}
	}
}
```
To be clear, I would instead like this output:
```php
<?php

class Foo
{
	public function bar($baz)
	{
		if ($baz == 900)
		{
			try
			{
				echo "Hello!";
			}
			catch (\Throwable $thrown)
			{
				echo "Error: {$thrown->getMessage()}";
			}
			finally
			{
				echo "Moving on.";
			}
		}

		if ($baz == true)
		{
			echo "Why?";
		}
		else
		{
			echo "Ha?";
		}
	}
}
```
I don't understand why the use of `braces` is not achieving this, though I guess I'm probably using it wrong based on all of my bug reports so far 😁 

Any advice to achieve that desired output would be appreciated. Please also let me know if I should raise this as a separate bug report instead.

**Comment by kubawerlos** [comment]
That was fixed with this integration test: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.18/tests/Fixtures/Integration/priority/single_space_after_construct%2Cbraces.test

Unfortunately, it's still waiting to be released - it's only affecting `2.18.4` as it changed significantly priority of the ~monster~ `braces`.

**Comment by Andrew-Staves-Activ** [comment]
Ok, thanks, when you mentioned that test it appeared to be just a test. I didn't realise that would actually ensure the fixers were applied in the order I'm after. I'm (obviously) not familiar enough with the internals of this to understand how the test is used to do that - but I'm very happy to take your word that it is, thank you.

I look forward to, I guess, `2.18.5` 😄

**Comment by kubawerlos** [comment]
You are actually interested in the diff that is in the test, the bottom version is given and upper is expected - and that will handle as you want it to.

**Comment by Andrew-Staves-Activ** [comment]
Sure, yes, I saw that the diff was what I wanted, it wasn't clear to me how having the test would ensure that this happened, but as I said, I'm happy to take your word that having it does achieve that. Thanks again.

**Comment by Wimble84** [comment]
That was fixed on 2.18.5 ! Thanks
